python/ray/tune/trainable.py

```diff
@@ -261,7 +261,8 @@ def save(self, checkpoint_dir=None):
             raise ValueError(
                 "`_save` must return a dict or string type: {}".format(
                     str(type(checkpoint))))
-        with open(checkpoint_path + ".tune_metadata", "wb") as f:
+        metadata_path = os.path.join(checkpoint_path, ".tune_metadata")
```

Comment:
this should be `checkpoint_dir` right?

Reply:
`checkpoint_path` is/can be the output of `self._save` which can be sort of arbitrary.

Reply:
I think the problem is that the `restore` method currently only knows about `checkpoint_path`. Should that be changed to take in a `checkpoint_dir` as well?

Reply:
Oh hm, one option also is to make _save return either a state dict or None, and in the None case, _restore will receive the checkpoint_dir again.

Right now, we have `checkpoint_dir = _save(); _restore(checkpoint_dir)`, but we don't necessarily need to have that structure.

Reply:
Oh I now see the problem and confusion. We currently expect the metadata to be at `.../[checkpoint_dir]/checkpoint.metadata` and not at `.../[checkpoint_dir]/checkpoint/.metadata`. I've only been using the string version of save/restore and thought that the `checkpoint_path` should point to a directory so that the metadata file becomes `...[checkpoint_path]/.metadata`. This works if the `_save` returns a path ending in `/`, but doesn't seem to be the expected behavior.

That means that this is currently working as expected and my change would break things.

Maybe the string returned by `_save` should always point to a directory instead of a file?

Reply:
That's one option, not clear what the tradeoffs are though... Related discussion in #4728